pygama/dsp/_processors/template.py

```diff
@@ -0,0 +1,77 @@
+# 1) Import python modules
+#
+import numpy as np
+from numba import guvectorize
+
+# 2) Provide instructions to numba
+# 
+# Documentation about numba guvectorize decorator:
+# https://numba.pydata.org/numba-doc/latest/user/vectorize.html#the-guvectorize-decorator
+#
+# Notes:
+# - use `nopython=True` and `cache=True`         (@ian, is this correct?)
+# - use 64-bit variables, e.g. int64 and float64 (@all, this is up for discussion)
```

Comment:
There are performance benefits to using 32-bit variables for processors that take advantage of SIMD vectorized operations, since more 32-bit values than 64-bit values can be processed at once

Reply:
if there is a clear advantage, I am fine with declaring multiple set of input variables. However, I would try to keep the number of sets limited, and define a clear standard on which sets should be declared. It is easy to have one set of 32-bit variables and another with 64-bit variables. I would however avoid sets with a mixture of 32- and 64-bit variables as it can be confusing to decide which is what.

Btw, which declaration is used when multiple are defined? I don't like that the user does not have control over this.

Reply:
Yeah, I agree that we should have some specification that prevents the number of signatures from blowing up. For waveforms and amplitude values, I think dividing between 32 and 64 bit makes sense. For timing values, I agree that floats are preferable, but there is (at least) one case where we will be getting int32s, which is if we use the argmax function from numpy.

When multiple declarations are available, it will use the types that match what is input. When calling it directly, it will cast values as needed, if the direct types aren't available; in the ProcessingChain class, I included code to find the right type signature. This was necessary so that the code can deduce how to create new internal memory buffers of the right type and size. If it encounters data types that are not compatible with any of the signatures, it will raise an exception. Unfortunately, the way I wrote it does not allow for casting to other types, so if it sees an int32, it won't cast up to float64 automatically.

Reply:
I added two declarations, one for 32-bit variables and the other one for 64-bit variables. I propose not to support the usage of integers as we need the NaN value to handle the expectations. Would it be possible to implement a check in the processing chain and cast any integer into a floa?